File: frontend/src/services/budget.service.ts
import { apiClient } from '../api/client';
import { Budget, CreateBudgetDTO } from '../types/budget';

export const BudgetService = {
  async getBudgets(): Promise<Budget[]> {
    const data = await apiClient<any[]>('/budgets');
    return (data || []).map((item) => ({
      id: item.id,
      user_id: item.userId || '',
      category_id: item.categoryId || '',
      category_name: item.categoryName || 'General Category',
      amount: Number(item.budgetAmount || item.amount),
      spent: Number(item.spentAmount || 0),
      remaining: Number(item.remainingAmount || 0),
      percentage: Number(item.percentageUsed || 0),
      month: Number(item.month),
      year: Number(item.year),
      created_at: item.createdAt,
    }));
  },

  async addBudget(dto: CreateBudgetDTO): Promise<Budget> {
    const data = await apiClient<any>('/budgets', {
      method: 'POST',
      body: JSON.stringify({
        categoryId: dto.category_id || 'cat-general',
        amount: dto.amount,
        month: dto.month,
        year: dto.year,
      }),
    });

    return {
      id: data.id,
      user_id: data.userId || '',
      category_id: data.categoryId || '',
      category_name: data.categoryName || dto.category_name || 'General Category',
      amount: Number(data.budgetAmount || data.amount),
      spent: Number(data.spentAmount || 0),
      remaining: Number(data.remainingAmount || 0),
      percentage: Number(data.percentageUsed || 0),
      month: Number(data.month),
      year: Number(data.year),
      created_at: data.createdAt,
    };
  },

  async deleteBudget(id: string): Promise<void> {
    await apiClient(`/budgets/${id}`, { method: 'DELETE' });
  },
};

File: backend/src/main/java/com/expensetracker/service/BudgetService.java
package com.expensetracker.service;

import com.expensetracker.dto.BudgetRequest;
import com.expensetracker.dto.BudgetResponse;
import com.expensetracker.exception.ResourceNotFoundException;
import com.expensetracker.model.Budget;
import com.expensetracker.model.Category;
import com.expensetracker.model.Expense;
import com.expensetracker.model.User;
import com.expensetracker.repository.BudgetRepository;
import com.expensetracker.repository.CategoryRepository;
import com.expensetracker.repository.ExpenseRepository;
import com.expensetracker.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class BudgetService {

    private final BudgetRepository budgetRepository;
    private final CategoryRepository categoryRepository;
    private final ExpenseRepository expenseRepository;
    private final UserRepository userRepository;

    @Transactional(readOnly = true)
    public List<BudgetResponse> getAllBudgets(String userId) {
        return budgetRepository.findByUserId(userId)
                .stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    @Transactional
    public BudgetResponse createOrUpdateBudget(String userId, BudgetRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        Category category = categoryRepository.findById(request.getCategoryId())
                .orElseThrow(() -> new ResourceNotFoundException("Category not found with id: " + request.getCategoryId()));

        Budget budget = budgetRepository.findByUserIdAndCategoryIdAndMonthAndYear(
                userId, request.getCategoryId(), request.getMonth(), request.getYear()
        ).orElseGet(() -> Budget.builder()
                .user(user)
                .category(category)
                .month(request.getMonth())
                .year(request.getYear())
                .build());

        budget.setAmount(request.getAmount());
        Budget saved = budgetRepository.save(budget);

        return mapToResponse(saved);
    }

    @Transactional
    public void deleteBudget(String userId, String id) {
        Budget budget = budgetRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Budget not found with id: " + id));

        if (!budget.getUser().getId().equals(userId)) {
            throw new IllegalArgumentException("Unauthorized access to resource");
        }

        budgetRepository.delete(budget);
    }

    public BudgetResponse mapToResponse(Budget budget) {
        String userId = budget.getUser().getId();
        LocalDate startDate = LocalDate.of(budget.getYear(), budget.getMonth(), 1);
        LocalDate endDate = startDate.plusMonths(1).minusDays(1);

        BigDecimal spent = expenseRepository.findByUserIdAndTransactionDateBetween(userId, startDate, endDate)
                .stream()
                .filter(e -> e.getCategory() != null && e.getCategory().getId().equals(budget.getCategory().getId()))
                .map(Expense::getAmount)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        BigDecimal remaining = budget.getAmount().subtract(spent);

        double percentage = budget.getAmount().compareTo(BigDecimal.ZERO) > 0
                ? spent.divide(budget.getAmount(), 4, RoundingMode.HALF_UP).doubleValue() * 100
                : 0.0;

        String status = "Good";
        if (percentage >= 90.0) {
            status = "Exceeded";
        } else if (percentage >= 70.0) {
            status = "Warning";
        }

        return BudgetResponse.builder()
                .id(budget.getId())
                .categoryId(budget.getCategory().getId())
                .categoryName(budget.getCategory().getName())
                .categoryIcon(budget.getCategory().getIcon())
                .categoryColor(budget.getCategory().getColor())
                .budgetAmount(budget.getAmount())
                .spentAmount(spent)
                .remainingAmount(remaining)
                .percentageUsed(Math.round(percentage * 100.0) / 100.0)
                .month(budget.getMonth())
                .year(budget.getYear())
                .status(status)
                .build();
    }
}
